File: ts-sdk/src/tasks.ts
import type { CallContext, LLMRequest, LLMResponse } from "./types.js";

export interface Task {
  before_call?: (request: LLMRequest, ctx: CallContext) => LLMRequest | void | Promise<LLMRequest | void>;
  after_call?: (
    request: LLMRequest,
    response: LLMResponse,
    ctx: CallContext,
  ) => LLMResponse | void | Promise<LLMResponse | void>;
}

File: python-sdk/axon/tasks.py
from __future__ import annotations

from typing import Awaitable, Protocol

from axon.types import CallContext, LLMRequest, LLMResponse


class Task(Protocol):
    def before_call(
        self, request: LLMRequest, ctx: CallContext
    ) -> LLMRequest | None | Awaitable[LLMRequest | None]: ...

    def after_call(
        self, request: LLMRequest, response: LLMResponse, ctx: CallContext
    ) -> LLMResponse | None | Awaitable[LLMResponse | None]: ...
